extmod/moducryptolib.c

```diff
@@ -356,9 +361,190 @@ STATIC const mp_obj_type_t ucryptolib_aes_type = {
     .locals_dict = (void *)&ucryptolib_aes_locals_dict,
 };
 
+#if MICROPY_PY_UCRYPTOLIB_GCM
+
+#if MICROPY_SSL_AXTLS
+#warning GCM is not currentl implementing with axTLS
+#endif
+
+#if MICROPY_SSL_MBEDTLS
+typedef struct _mp_obj_aesgcm_t {
+    mp_obj_base_t base;
+    int key_len;
+    uint8_t key[32];
+} mp_obj_aesgcm_t;
+
+STATIC mp_obj_t ucryptolib_aesgcm_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+
+    mp_obj_aesgcm_t *o = m_new_obj(mp_obj_aesgcm_t);
+
+    o->base.type = type;
+    mp_buffer_info_t keyinfo;
+    mp_get_buffer_raise(args[0], &keyinfo, MP_BUFFER_READ);
+    if (32 != keyinfo.len && 16 != keyinfo.len) {
+        mp_raise_ValueError(MP_ERROR_TEXT("key length"));
+    }
+
+    o->key_len = keyinfo.len;
+    memcpy(&o->key, keyinfo.buf, keyinfo.len);
+
+    return MP_OBJ_FROM_PTR(o);
+}
+
+STATIC mp_obj_t ucryptolib_aesgcm_enc_dec_check_args(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs,
+    mp_buffer_info_t *nonce, mp_buffer_info_t *input, mp_buffer_info_t *output, mp_buffer_info_t *ad, int in_out_delta) {
+    mp_obj_t out_obj;
+    mp_map_elem_t *map_element;
+
+    if (n_args != 3) {
+        mp_raise_ValueError(MP_ERROR_TEXT("2 positional arguments required"));
+    }
+
+    mp_get_buffer_raise(args[1], nonce, MP_BUFFER_READ);
+    mp_get_buffer_raise(args[2], input, MP_BUFFER_READ);
+
+    size_t output_len = input->len + in_out_delta;
+
+    map_element = mp_map_lookup(kwargs, MP_OBJ_NEW_QSTR(MP_QSTR_out_buf), MP_MAP_LOOKUP);
+    if (map_element != NULL) {
+        out_obj = map_element->value;
+        mp_get_buffer_raise(out_obj, output, MP_BUFFER_WRITE);
+        if (output->len != output_len) {
+            mp_raise_ValueError(MP_ERROR_TEXT("incorrect output length"));
+        }
+    } else {
+        out_obj = MP_OBJ_NULL;
+        output->buf = m_new(char, output_len);
+        output->len = output_len;
+    }
+
+    map_element = mp_map_lookup(kwargs, MP_OBJ_NEW_QSTR(MP_QSTR_adata), MP_MAP_LOOKUP);
+    if (map_element != NULL) {
+        mp_get_buffer_raise(map_element->value, ad, MP_BUFFER_READ);
+    } else {
+        ad->buf = NULL;
+        ad->len = 0;
+    }
+
+    // Maybe this should check for spurious keyword args...
+
+    return out_obj;
+}
+
+#define UCRYPTOLIB_AESGCM_TAG_LEN 16
+
+STATIC mp_obj_t ucryptolib_aesgcm_encrypt(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs) {
+    mp_obj_aesgcm_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_buffer_info_t nonce;
+    mp_buffer_info_t input;
+    mp_buffer_info_t output;
+    mp_obj_t out_obj;
+    mp_buffer_info_t ad;
+
+    out_obj = ucryptolib_aesgcm_enc_dec_check_args(n_args, args, kwargs,
+        &nonce, &input, &output, &ad,
+        UCRYPTOLIB_AESGCM_TAG_LEN);
+
+    mbedtls_gcm_context gcm;
+    int rc;
+
+    mbedtls_gcm_init(&gcm);
+
+    rc = mbedtls_gcm_setkey(&gcm, MBEDTLS_CIPHER_ID_AES, self->key, self->key_len * 8);
```

Comment:
I was going to do that, and then I looked at the size of the expanded key schedule and decided to lean the memory/cpu trade-off in favour of less memory. Also, unlike the other AES cipher modes, this does not implement incremental encryption and decryption; you have to do the operation on all the data in one shot, so the benefit of keeping an expanded key around is more limited. If we wanted to support incremental operation we would need a whole new API since there would need to be an explicit finalise process to compute/check the authentication tag. Given the way most people use AESGCM I decided to keep the API simple. Furthermore with mbedTLS the expansion for encryption and decryption are different, so either I'd have to keep copies or you'd need different objects for each direction. Unless there is evidence that the key expansion process is a real performance bottleneck I'd rather leave it small and simple.

Reply:
I saw indeed that the expanded key size is > 300 bytes, which is also considerable chunk of the stack, this is a tradeoff that should be considered.

About the expanded key: AFAIK, GCM being based on a counter mode means that it uses the "encryption" key for both sides (can also that `mbedtls_gcm_setkey` calls the cipher's `setkey` with `MBEDTLS_ENCRYPT`).

About the API: I see what you're saying, perhaps consider separating to `update` and `finalize` to allow incrementality (following the API of `hashlib`). Anyway even if it's not supported, reusing the expanded key for each encrypted/decrypted message may be beneficial.
I don't have any metrics on the key expansion, it's one of those things you just "know" and try to follow, if possible; whenever I'd needed to use AES encryption I tried to reuse the encryptor object if possible. So it's nicer to create an API that allows it :) It's indeed a tradeoff of stack size / heap size then. Should consider / document it, at least.

Reply:
Well, if you want I could add a `#define` to let people decide if they want to use the CPU or the stack space. For what it's worth, having spend the last 25 years in the cryptography business I've come to know that you should never blindly trust _'those things you just "know"'_.

A multi-step API for AESGCM would require separate `start`, `update` and `finalise` functions since (as with any sort of counter mode) it's critical to use a different nonce for each message, and that would need to be passed in at the start. The method signature for `finalise` would also need to be different for encrypt and decrypt operations, since in the former the tag is an output while in the later the tag is an input. In MicroPython we could fake that since we'd want to offer the option to pass in a writable buffer to the encrypt as a destination for the tag and the implementation's C function return type for `None` is still an object. For users who have large volumes of data to encrypt and they either are generating the input incrementally or they plan to write the encrypted data out to the network or storage incrementally, there are advantages to the incremental API. The question is are those use cases common enough to warrant a substantially more complex API?

Reply:
It's worth noting that the CPython `cryptography` library offers [ `AESGCM`](https://cryptography.io/en/latest/hazmat/primitives/aead/#cryptography.hazmat.primitives.ciphers.aead.AESGCM) but does _not_ offer an incremental API for that. Since the method signatures for `encrypt` and `decrypt` here are now compatible with the `cryptography` library I don't think that we need to add another, more complex incremental API.

Reply:
Like you pointed in the previous comment, my thoughts behind the incremental API were for those generating incremental data. And, well, CPython libraries may not provide incremental API, but they are not memory-constrained like most MicroPython platforms.

I'm not sure those use cases are common enough, let's stay with the `encrypt`/`decsrypt` only.